brain/morphology.cpp

```diff
@@ -0,0 +1,258 @@
+
+/* Copyright (c) 2013-2015, EPFL/Blue Brain Project
+ *                          Juan Hernando <jhernando@fi.upm.es>
+ *
+ * This file is part of Brion <https://github.com/BlueBrain/Brion>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "morphology.h"
+
+#include <brion/morphology.h>
+#include <brion/circuit.h>
+
+#include <lunchbox/log.h>
+
+#include <bitset>
+
+#include <math.h>
+
+namespace brain
+{
+
+namespace
+{
+    void nullDeleter( const brion::Morphology* ) {}
+}
+
+class Morphology::Impl
+{
+public:
+    boost::shared_ptr< const brion::Morphology > _morphology;
```

Comment:
This is mainly motivated by the prototype of brion::morphologies::load (which is still RFC). This set of functions returns std::vector< boost::shared_ptr<brion::Morphology> > because brion::Morphology is non copyable and we can't use movables from C++11 (in case CoreNeuron finally adopts it, which implies Brion has to compile in BG).

With this shared_ptr you can have the 3 possibilities provided by the brain::Morphology constructors:
* Create a brion::Morphology right away
* Take a shared pointer
* Take a reference (which is stored as a shallow shared_ptr with null deallocator)
I could also make brain::Morphology always take a const brion::Morphology* to imply that the constructor will store a reference, but that puts the whole responsibility of managing the memory on the caller and I find it less convenient.